source/common/stats/stats_impl.h

```diff
@@ -366,9 +396,135 @@ class HistogramImpl : public Histogram, public MetricImpl {
   // Stats::Histogram
   void recordValue(uint64_t value) override { parent_.deliverHistogramToSinks(*this, value); }
 
+  void merge() override {}
+
+  bool used() const override { return true; }
+
+  HistogramStatistics intervalStatistics() const override { return HistogramStatistics(); }
+
+  HistogramStatistics cumulativeStatistics() const override { return HistogramStatistics(); }
+
   Store& parent_;
 };
 
+/**
+ * Log Linear Histogram implementation per thread.
+ */
+class ThreadLocalHistogramImpl : public Histogram, public MetricImpl {
+public:
+  ThreadLocalHistogramImpl(const std::string& name, Store& parent, std::string&& tag_extracted_name,
+                           std::vector<Tag>&& tags)
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {
+    histograms_[0] = hist_alloc();
+    histograms_[1] = hist_alloc();
+    current_active_ = 0;
+    flags_ = 0;
+  }
+
+  ~ThreadLocalHistogramImpl() {
+    hist_free(histograms_[0]);
+    hist_free(histograms_[1]);
+  }
+  // Stats::Histogram
+  void recordValue(uint64_t value) override {
+    hist_insert_intscale(histograms_[current_active_], value, 0, 1);
+    parent_.deliverHistogramToSinks(*this, value);
+    flags_ |= Flags::Used;
+  }
+
+  void beginMerge() { current_active_ = 1 - current_active_; }
+
+  void merge() override {}
+
+  bool used() const override { return flags_ & Flags::Used; }
+
+  HistogramStatistics intervalStatistics() const override { return HistogramStatistics(); }
+
+  HistogramStatistics cumulativeStatistics() const override { return HistogramStatistics(); }
+
+  void merge(histogram_t* target) {
+    histogram_t* hist_array[1];
+    hist_array[0] = histograms_[1 - current_active_];
+    hist_accumulate(target, hist_array, ARRAY_SIZE(hist_array));
+    hist_clear(hist_array[0]);
+  }
+
+  Store& parent_;
+  int current_active_;
+  histogram_t* histograms_[2];
+  std::atomic<uint16_t> flags_;
+};
+
+typedef std::shared_ptr<ThreadLocalHistogramImpl> TlsHistogramSharedPtr;
+
+/**
+ * Log Linear Histogram implementation that is stored in the main thread.
+ */
+class HistogramParentImpl : public Histogram, public MetricImpl {
+public:
+  HistogramParentImpl(const std::string& name, Store& parent, std::string&& tag_extracted_name,
+                      std::vector<Tag>&& tags)
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {
+    interval_histogram_ = hist_alloc();
+    cumulative_histogram_ = hist_alloc();
+  }
+
+  ~HistogramParentImpl() {
+    hist_free(interval_histogram_);
+    hist_free(cumulative_histogram_);
+  }
+
+  // Stats::Histogram
+  void recordValue(uint64_t) override {}
+
+  bool used() const override {
+    bool any_tls_used = false;
+    for (auto tls_histogram : tls_histograms_) {
```

Comment:
All sues of tls_histograms_ have to be locked if we are going to do the used() stuff. I think what I would actually do is move your "merge_lock" into each histogram. and then just lock each function here that is needed. Let's go ahead and do that and I can review again. (Just trying to iteratively tighten stuff up so I can focus more on concurrency).

Reply:
ok. Moved the merge_lock to histogram and locked the functions there. Looks like this is much cleaner.